src/pages/index.js

```diff
@@ -63,27 +102,110 @@ function handleImageClick(cardLink, cardCaption) {
   imagePopup.open(cardLink, cardCaption);
 }
 
+function handleAvatarClick() {
+  editAvatarPopup.open();
+}
+
+function handleDeletePopup(cardId, evt) {
+  deleteVerifyPopup.open();
+  cardIdForDeletion = cardId;
+  cardToDelete = evt.target.closest(".card");
+}
+
+function likeToServer(cardId, cardElement) {
+  if (!cardElement.querySelector(".card__liked_active")) {
+    api.removeLikeFromCard(cardId).catch((err) => console.error(err));
+  } else {
+    api.addLikeToCard(cardId).catch((err) => console.error(err));
+  }
+}
+
 function createCard(imageInfo) {
-  const newCard = new Card(imageInfo, "#card-template", handleImageClick);
+  const newCard = new Card(
+    imageInfo,
+    "#card-template",
+    handleImageClick,
+    handleDeletePopup,
+    likeToServer
+  );
   const cardElement = newCard.generateCard();
   return cardElement;
 }
 
+function renderLoading(isLoading, button) {
+  if (isLoading) {
+    buttonText = button.textContent;
+    button.textContent = "Saving...";
+  } else {
+    button.textContent = buttonText;
+  }
+}
+
 function handleAddImage(imageInfo) {
-  const addedCardElement = createCard(imageInfo);
-  cardSection.addItem(addedCardElement, "prepend");
-  addFormPopup.close();
-  addFormPopup.clearForm();
-  formValidators["add"].disableSubmitButton();
+  renderLoading(true, addPopup.querySelector(config.submitButtonSelector));
+  api
+    .addCard(imageInfo)
+    .then((res) => {
+      let addedCardElement = createCard(res);
+      cardSection.addItem(addedCardElement, "prepend");
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      renderLoading(false, addPopup.querySelector(config.submitButtonSelector));
+      addFormPopup.close();
+      addFormPopup.clearForm();
+      formValidators["add"].disableSubmitButton();
+    });
 }
 
 function handleEditProfile(newProfileInfo) {
-  userInfo.setUserInfo(newProfileInfo);
-  editFormPopup.close();
+  renderLoading(true, editPopup.querySelector(config.submitButtonSelector));
+  userInfo.changeUserInfo(newProfileInfo);
+  api
+    .editUserInfo(newProfileInfo)
+    .then(() => userInfo.setUserInfo(profileInfo))
+    .catch((err) => console.error(err))
+    .finally(() => {
+      renderLoading(
+        false,
+        editPopup.querySelector(config.submitButtonSelector)
+      );
+      editFormPopup.close();
+    });
+}
+
+function handleChangeAvatar(newAvatarImage) {
+  renderLoading(true, avatarPopup.querySelector(config.submitButtonSelector));
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```